torch/csrc/jit/codegen/cuda/ir_builder.cpp

```diff
@@ -20,9 +22,13 @@ Val* IrBuilder::newScalar(DataType dtype) {
       return IrBuilder::create<Double>(dtype);
     case DataType::Int:
     case DataType::Int32:
+    case DataType::Index:
```

Comment:
Should Index not return Index type since it's a special type?

Reply:
This is complicated. I don't think we have a consistent protocol on which should be used. See the comment in `newArithmeticExpr`:

```C++
  // In principle, we should keep these IrBuilder functions as
  // simple as possible since they are just used by the lowering for
  // scalar computations. We should enforce strict typing with no
  // implicit type promotion unless required. However, for
  // int and int64_t, our usages are pretty loose in many places. Originally we
  // only had int64_t, then we added nvfuser_index_t and replaced the types of
  // some of the values from int64_t to int just at the beginning of lowering.
  // This resulted in inconsistent usages of integer types in many places, and
  // fixing all of them to make everything consistent would be a lot of work
  // than just allowing the integer type promotion for the two inputs as below.
  // Note that this is only needed for integer types. See also PR #2228.
```

Reply:
A bit confused. When `dtype` is `DataType::Index`, this should return a `DataType::Index` Int. Note that we only have one class for all variants of integer types.

Reply:
I agree. Just a bit annoying that we are not using `DataType::Index` everywhere when possible.